### PR TITLE
Limit broadcast-based equivalence to most recent broadcasts

### PR DESCRIPTION
**Pull request by tristanbeardmore-gray**
Efficiency improvement to avoid issues with certain content that can have thousands of broadcasts.

Lesser-used generators haven't been changed for now, partially due to time and due to them being lower risk.

### CODE REVIEW

File: src/main/java/org/atlasapi/equiv/generators/barb/BarbBroadcastMatchingItemEquivalenceGeneratorAndScorer.java [hunk 1]
```diff
@@ -113,35 +121,55 @@ public static Builder builder() {
...
+        Stream<Broadcast> mostRecentBroadcasts = subject.getVersions().stream()
+                .map(Version::getBroadcasts)
+                .flatMap(Collection::stream)
+                .filter(Broadcast::isActivelyPublished)
+                .filter(broadcastFilter)
+                .sorted(MOST_RECENT_BROADCAST_COMPARATOR)
+                .distinct();
+        if (numberOfMostRecentBroadcastsToCheck != null) {
+            mostRecentBroadcasts = mostRecentBroadcasts.limit(numberOfMostRecentBroadcastsToCheck);
+        }
```
Comment (pawel-mtbcst) [inline src/main/java/org/atlasapi/equiv/generators/barb/BarbBroadcastMatchingItemEquivalenceGeneratorAndScorer.java:169]: `.limit(firstNonNull(numberOfMostRecentBroadcastsToCheck, MAX_VALUE))`

Long.MAX_VALUE
Objects.numberOfMostRecentBroadcastsToCheck

File: src/main/java/org/atlasapi/equiv/update/updaters/providers/item/barb/BbcTxlogsItemUpdaterProvider.java [hunk 2]
```diff
@@ -79,6 +78,7 @@ public static BbcTxlogsItemUpdaterProvider create(boolean isSubjectTxlog) {
                                                         .withContainerCacheDuration(60)
                                                         .build()
                                         )
+                                        .withNumberOfMostRecentBroadcastsToCheck(50)
```
Comment (pawel-mtbcst) [inline src/main/java/org/atlasapi/equiv/update/updaters/providers/item/barb/BbcTxlogsItemUpdaterProvider.java:81]: Looks like it's something we may want to change in the future.

File: src/main/java/org/atlasapi/equiv/generators/StandardBroadcastEquivalenceGeneratorAndScorer.java [hunk 3]
```diff
@@ -174,6 +185,24 @@ public static Builder builder() {
...
+            Predicate<? super Broadcast> broadcastFilter
+    ) {
+        Stream<Broadcast> mostRecentBroadcasts = subject.getVersions().stream()
+                .map(Version::getBroadcasts)
+                .flatMap(Collection::stream)
+                .filter(Broadcast::isActivelyPublished)
+                .filter(broadcastFilter)
+                .sorted(MOST_RECENT_BROADCAST_COMPARATOR)
+                .distinct();
+        if (numberOfMostRecentBroadcastsToCheck != null) {
```
[the remark this answers is not in the document]
Comment (tristanbeardmore-gray) [reply src/main/java/org/atlasapi/equiv/generators/StandardBroadcastEquivalenceGeneratorAndScorer.java:200]: This ended up being ugly given types where the first is Integer and the second is Long; firstNonNull didn't want to play ball without additional code.

[the remark this answers is not in the document]
Comment (pawel-mtbcst) [reply src/main/java/org/atlasapi/equiv/generators/StandardBroadcastEquivalenceGeneratorAndScorer.java:200]: @tristanbeardmore-gray using Integer.MAX_VALUE works:
image

[the remark this answers is not in the document]
Comment (pawel-mtbcst) [reply src/main/java/org/atlasapi/equiv/generators/StandardBroadcastEquivalenceGeneratorAndScorer.java:200]: Or casting the var to long if we're concerned about the MAX_VALUE of intger. This is still prettier than if :)
`.limit(firstNonNull((long)limit, Long.MAX_VALUE))`
image

[the remark this answers is not in the document]
Comment (pawel-mtbcst) [reply src/main/java/org/atlasapi/equiv/generators/StandardBroadcastEquivalenceGeneratorAndScorer.java:200]: ofc if we accepted lowering the limit

[the remark this answers is not in the document]
Comment (pawel-mtbcst) [reply src/main/java/org/atlasapi/equiv/generators/StandardBroadcastEquivalenceGeneratorAndScorer.java:200]: ofc if we accepted lowering the limit
If not, you can use conditional operator, which is still prettier :) 
`.limit(limit != null ? limit : Long.MAX_VALUE)`

**Pull request merged**